### DISCUSSION TITLE
How do I specify an interpreter during build?

### DISCUSSION THREAD

**Discussion by pmeier** [PyO3/maturin#1357] [Q&A] [answered]
Either I'm to dense to find it, but so far I have no idea what I need to pass to `maturin build --interpreter=???`. Help message is not really helpful:

```
$ maturin build --help | grep -A1 -e --interpreter
    -i, --interpreter <INTERPRETER>...
            The python versions to build wheels for, given as the names of the interpreters
```

I tried the "raw" specifiers such as `"3.8"`, but for that I'm getting

```
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
🐍 Not using a specific python interpreter
[...]
📦 Built wheel for abi3 Python ≥ 3.7 [...]
```

Using the tags from [PEP 427](https://peps.python.org/pep-0427/#file-name-convention), i.e. `"py38"`, `"cp38"` fails:

```
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
💥 maturin failed
  Caused by: The given list of python interpreters is invalid
  Caused by: Python interpreter `py38` doesn't exist
```

From the failure message it seems I should also be able to build multiple wheels at the same time by passing a list of interpreters. Since the "raw" specifier was the only one that didn't fail, I tried a comma-separated list:

```
$ maturin build --interpreter=3.8,3.9
[...]
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
💥 maturin failed
  Caused by: The given list of python interpreters is invalid
  Caused by: Python interpreter `3.8,3.9` doesn't exist
```

What am I missing here? Happy to contribute to improve the help message or documentation.

**Comment by konstin** [comment, marked as answer]
I didn't realize we wrote it that ambiguously, it's not actually the name of the interpreter, but the executable like you would in the shell, such as `python3.7` or `/usr/bin/python3.8`

**Reply by pmeier** [reply]
That doesn't seem to work for me as well:

```
$ python -V
Python 3.7.15
$ maturin build -i python3.7
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
🐍 Not using a specific python interpreter
[...]
$ maturin build -i python
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
🐍 Not using a specific python interpreter
[...]
$ maturin build -i $(which python)
🍹 Building a mixed python/rust project
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
🐍 Not using a specific python interpreter
[...]
```

**Reply by konstin** [reply]
Oh sorry, it seems you're the `abi3` feature of pyo3. In that case no need for using any specific python interpreter, the wheel compatibility is determined by pyo3 features; neither pyo3 nor maturin do even need to query or link python in that case.

**Reply by pmeier** [reply]
> Oh sorry, it seems you're the `abi3` feature of pyo3.

Yes, i have

```toml
[dependencies.pyo3]
version = "0.17"
features = ["extension-module", "abi3-py37"]
```

in my `Cargo.toml`.

> In that case no need for using any specific python interpreter, the wheel compatibility is determined by pyo3 features; neither pyo3 nor maturin do even need to query or link python in that case.

Ok, how can I get the "proper" wheel tags in that case? Right now every wheel will be `XXX-cp37-abi3-YYY`. AFAIK, `pip` will reject that for every other Python version. Do I copy the wheel and just `sed` the Python version? Do I `sed` the `"abi-py37"` in `Cargo.toml`?

**Reply by konstin** [reply]
nope, those wheels can be installed on any later python 3 version! abi3 is really cool because it allows you to build only one wheel for all python versions you want to support

**Reply by messense** [reply]
FYI, here is the `abi3` docs in pyo3 user guide: https://pyo3.rs/v0.17.3/building_and_distribution.html#py_limited_apiabi3

**Reply by pmeier** [reply]
Oh nice, I didn't know about that. Thanks!
